---
title: "SCTransform, clustering, marker genes"
author: "Friederike DÃ¼ndar"
date: "6/27/2019; updated `r Sys.Date()`"
output:
    html_document:
        toc: true
        toc_depth: 6
        code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE, cache.lazy = FALSE, warning = FALSE, eval=FALSE)
```

```{r}
library(scater)
library(Seurat) # v. 3.1.0.
library(sctransform) # sctransform_0.2.0

load(file = "/scratchLocal/frd2007/2019_06_Jingli/data/sce_CellGeneFilt_2019-06.rda")
```

```{r seurat_processing, eval=FALSE}
## convert to Seurat object
srt <- CreateSeuratObject(counts = counts(sce.filt), 
                          project = "Epi", assay = "RNA",
                          min.cells = 0, min.features = 0,
                          names.field = 1#, 
                         # meta.data = as.data.frame(colData(sce.filt) )
)

srt <- SCTransform(srt, new.assay.name = "scTrans")
```

`SCTransform`  returns a Seurat object with a new assay and:
 - `counts` being (corrected) counts, i.e., Pearson residuals are turned into counts (as if all cells had the same number of UMI) [Ref](https://github.com/ChristophH/sctransform/issues/17)
 - `data` being log1p(corrected counts) --> this is what's going to be used for visualizations etc.!
 - `scale.data` being pearson residuals
    - "The residuals of our 'regularized negative binomial regression' represent effectively normalized data values that are no longer influenced by technical characteristics, but preserve heterogeneity driven by distinct biological states."
    - "Pearson residuals are independent of sequencing depth, and can be used for *variable gene selection, dimensional reduction, clustering, visualization, and differential expression*"
    - the Pearson residuals in the scale.data slot are the ones of the most variable genes!

### scTransform

```{r seurat_scTransform_PCA, eval=FALSE}
### PCA  ==========================================
srt <- RunPCA(object = srt,
              features = VariableFeatures(object = srt),
              do.print = FALSE)

png("elbow_plot_scTransform.png")
ElbowPlot(object = srt) # -->  to determine the # of dims for tSNE
dev.off()

png("pca_dimLoadings_scTransform.png", height = 1900, width = 1500)
VizDimLoadings(srt, dims = 1:20, reduction = "pca")
dev.off()

png("pca_dimPlot_scTransform.png", height = 1500, width = 1500)
DimPlot(srt)
dev.off()

png("pca_dimHeatmaps_scTransform.png", height = 1500, width = 800)
DimHeatmap(srt, dims = 1:20,  balanced = TRUE) #equal number of genes with both + and - scores
dev.off()
```

#### UMAP

```{r seurat_scTransform_umap, eval=FALSE}
srt <- RunUMAP(object = srt, 
               reduction = "pca",
               dims = 1:15,# check the elbow plot
               n.neighbors = 30L,
               metric = "correlation")

png("umap_scTransform.png")
DimPlot(srt, reduction = "umap")
dev.off()
```
```{r clustering_sctransform, eval=FALSE}
### Clusters =======================================
srt <- FindNeighbors(srt, dims = 1:15) #construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity)
srt <- FindClusters(srt, resolution = .8) #Louvain algorithm 
srt <- FindClusters(srt, resolution = .3) #Louvain algorithm 

##!save(srt, file = "/scratchLocal/frd2007/2019_06_Jingli/data/seurat_scTransform.rda")

#markers_seurClsts <- FindAllMarkers(srt)
#save(markers_seurClsts, file = "/athena/abc/scratch/frd2007/2019-05_MikeNotaras/data/markers_seurat_scTransform_June21.rda")


png("var_genes_scTransform.png")
plot1 <- VariableFeaturePlot(srt)
plot2 <- LabelPoints(plot = plot1, 
                     points = head(VariableFeatures(srt), 20), 
                     repel = TRUE)
print(plot2)
dev.off()
```

### Batch correction

```{r sctransform_with_batch_correction}
srt$condition <- gsub("-.*","",names(srt$nCount_RNA))

srt.batch <- SCTransform(srt, new.assay.name = "scTrans_batch", batch_var = 'condition')
### PCA  ==========================================
srt.batch <- RunPCA(object = srt.batch,
              features = VariableFeatures(object = srt.batch),
              do.print = FALSE)

png("elbow_plot_scTransformBatch.png")
ElbowPlot(object = srt.batch) # -->  to determine the # of dims for tSNE
dev.off()

png("pca_dimLoadings_scTransformBatch.png", height = 1900, width = 1500)
VizDimLoadings(srt.batch, dims = 1:25, reduction = "pca")
dev.off()

png("pca_dimPlot_scTransformBatch.png", height = 1500, width = 1500)
DimPlot(srt.batch, reduction = "pca",  group.by = "orig.ident")
dev.off()

png("pca_dimHeatmaps_scTransformBatch.png", height = 1500, width = 800)
DimHeatmap(srt.batch, dims = 1:25,  balanced = TRUE) #equal number of genes with both + and - scores
dev.off()

srt.batch <- RunUMAP(object = srt.batch, 
               reduction = "pca",
               dims = 1:30,# check the elbow plot
               n.neighbors = 30L,
               metric = "correlation")

png("umap_scTransformBatch.png")
DimPlot(srt.batch, reduction = "umap", group.by = "orig.ident")
dev.off()

### Clusters =======================================
srt.batch <- FindNeighbors(srt.batch, dims = 1:30) #construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity)
srt.batch <- FindClusters(srt.batch, resolution = .8) #Louvain algorithm 
srt.batch <- FindClusters(srt.batch, resolution = .1) #Louvain algorithm 
srt.batch <- FindClusters(srt.batch, resolution = .3) #Louvain algorithm 
markers_srt.batch.03 <- FindAllMarkers(srt.batch)


save(srt.batch, file = "/scratchLocal/frd2007/2019_06_Jingli/data/seurat_scTransformBatchCorrected.rda")

save(markers_srt.batch.03, file = "/scratchLocal/frd2007/2019_06_Jingli/data/markers_seurat_res03.rda")
```


```{r seurat_scTransform_combine_with_sce, eval=FALSE}
## Add Seurat info back to SCE ==========================================

### clustering
cd_srt <-  srt.batch@meta.data
sce.filt$res.03 <- cd_srt[colnames(sce.filt),]$scTrans_batch_snn_res.0.3
sce.filt$res.08 <- cd_srt[colnames(sce.filt),]$scTrans_batch_snn_res.0.8

### reduced dimensions
reducedDim(sce.filt, "pca_scTransformBatch") <- Embeddings(object = srt.batch, reduction = "pca")[colnames(sce.filt),]
reducedDim(sce.filt, "umap_scTransformBatch") <- Embeddings(object = srt.batch, reduction = "umap")[colnames(sce.filt),]

### normalized data
rnms <- data.table(ori = row.names(sce.filt)) # get original rownames -- Seurat replaces underscores with hyphens...
rnms[,seurat := gsub("_","-", ori)]

## the default setting of sctransform::vst is to remove genes expressed in fewer than 5 cells,
## therefore I need to adjust the data set
normd <- as.data.table(as.matrix(GetAssayData(object = srt.batch)), keep.rownames = TRUE) # log1p(corrected counts, i.e. Pearson residuals)
setnames(normd, "rn", "seurat")
normd <- rnms[normd, on = "seurat"]
dim(sce.filt)
#[1] 24127 15141
sce.filt <- sce.filt[normd$ori, ] ## filtering!
dim(sce.filt)
#> [1] 20329 15141

normd.m <- as.matrix(normd[, -c("seurat", "ori"), with=FALSE])
rownames(normd.m) <- normd$ori
normd.m <- normd.m[rownames(sce.filt), colnames(sce.filt)]
assay(sce.filt, "log1p_sctransform.batch") <- normd.m

saveRDS(sce.filt, file = "/scratchLocal/frd2007/2019_06_Jingli/data/sce_CellGeneFiltWithScTransformBatchCorrect_2019-06.rds")

rm(normd)
gc()

```

### SC3-based marker genes

Using the clustering from Seurat (resolution: 0.3)

```{r eval=FALSE}
## SC3-based markers
## following prescription here: https://github.com/loosolab/cardiac-progenitors/blob/master/src/04_clustering.R
markers_sc3.batch_res03 <- SC3::get_marker_genes( assay(sce.filt, "log1p_sctransform.batch"),
                                                  as.numeric(sce.filt$res.03))
row.names(markers_sc3.batch_res03) <- rownames(sce.filt)
save(markers_sc3.batch_res03, file = "/scratchLocal/frd2007/2019_06_Jingli/data/markers_SC3_2019-07.rda")
```

```{r}
load("/Users/frd2007/Documents/Projects/2019-06_Jingli/data/markers_SC3_2019-07.rda") # markers_sc3.batch_res03

markers_sc3.batch_res03 <- subset(markers_sc3.batch_res03, !is.na(auroc))
markers_sc3.batch_res03$clusts <- factor(markers_sc3.batch_res03$clusts - 1, levels = levels(scf$res.03), ordered = TRUE)
```

```{r}
subset(markers_sc3.batch_res03, auroc > 0.7 & pvalue <= 0.01) %>% ggplot(., aes(x = clusts, fill = clusts)) + geom_bar() + scale_fill_manual(values = ABCutilities:::fx.get_palette_ABC("paired_pal")[1:length(unique(scf$res.03))])

```
